File: Asignaturas DAW/Despliegue de aplicaciones web/Tareas/Tarea 3/Cartas/src/cartas/Baraja.java
package cartas;

/**
 * Clase abstracta Baraja
 * @author Hadrián
 */
public abstract class Baraja implements BarajaCartas {

    /**
     * Array de objetos carta (Baraja)
     */
    protected Carta[] cartas;
    
    /**
     * Número de posición de la siguiente carta
     */
    protected int posSiguienteCarta = 0;
    
    /**
     * Número total de cartas
     */
    protected static int NUM_CARTAS = 40;
    
    /** 
     * Devuelve un array de objetos Carta
     * @return Carta[]
     */
    public Carta[] getCartas() {
        return cartas;
    }
    
    /** 
     * Devuelve la posición de la siguiente carta (número)
     * @return int
     */
    public int getPosSiguienteCarta() {
        return posSiguienteCarta;
    }
    
    /** 
     * Devuelve el número de cartas de la baraja
     * @return int
     */
    public static int getNUM_CARTAS() {
        return NUM_CARTAS;
    }
    
    /** 
     * Asigna un array de cartas a la baraja
     * @param cartas
     */
    public void setCartas(Carta[] cartas) {
        this.cartas = cartas;
    }
    
    /** 
     * Asigna la posición de la siguiente carta de la baraja
     * @param posSiguienteCarta
     */
    public void setPosSiguienteCarta(int posSiguienteCarta) {
        this.posSiguienteCarta = posSiguienteCarta;
    }
    
    /** 
     * Asigna el número de cartas de la baraja
     * @param NUM_CARTAS
     */
    public static void setNUM_CARTAS(int NUM_CARTAS) {
        NUM_CARTAS = NUM_CARTAS;
    }
    
    /** 
     * Añade una carta a la baraja
     * @param unaCarta
     */
    protected void addCarta(Carta unaCarta) {
        this.cartas[cartas.length - 1] = unaCarta;
    }
    
    /** 
     * Método abstracto que crea una baraja
     */
    abstract void crearBaraja();

    /**
     * Baraja las cartas
     */
    public void barajar() {
    }

    /**
     * Pasa a la siguiente carta
     */
    public void siguienteCarta() {
    }

    /**
     * Devuelve el número de cartas de la baraja
     */
    public void cartasDisponibles() {
    }

    /** 
     * Reparte un número de cartas
     * @param cartasPedidas
     */
    public void repartirCartas(int cartasPedidas) {
    }

    /**
     * Devuelve el número de cartas repartidas
     */
    public void cartasRepartidas() {
    }

    /**
     * Muestra toda la baraja
     */
    public void mostrarBaraja() {
    }
}

File: Asignaturas DAW/Entorno Servidor/EjemplosYTareas/Tarea4/Baraja.php
<?php

require_once './Carta.php';
require_once './BarajaCartas.php';

abstract class Baraja implements BarajaCartas {

    protected $cartas = array();
    protected $posSiguienteCarta = 0;
    protected static $NUM_CARTAS = 40;

    function __construct() {
        
    }

    function getCartas() {
        return $this->cartas;
    }

    function getPosSiguienteCarta() {
        return $this->posSiguienteCarta;
    }

    static function getNUM_CARTAS() {
        return self::$NUM_CARTAS;
    }

    function setCartas($cartas) {
        $this->cartas = $cartas;
    }

    function setPosSiguienteCarta($posSiguienteCarta) {
        $this->posSiguienteCarta = $posSiguienteCarta;
    }

    static function setNUM_CARTAS($NUM_CARTAS) {
        self::$NUM_CARTAS = $NUM_CARTAS;
    }

    protected function addCarta($unaCarta) {
        if ($unaCarta instanceof Carta) {
            $this->cartas[] = $unaCarta;
        }
    }
    
    abstract function crearBaraja();

    function barajar() {
        shuffle($this->cartas);
    }

    function siguienteCarta() {
        if ($this->posSiguienteCarta < self::$NUM_CARTAS) {
            echo $this->cartas[$this->posSiguienteCarta]->toString()."<br>";
            $this->posSiguienteCarta++;
        } else {
            return "No hay más cartas";
        }
    }

    function cartasDisponibles() {
        $numCartasDisponibles = self::$NUM_CARTAS - $this->posSiguienteCarta;
        echo "Se pueden repartir ".$numCartasDisponibles." cartas<br><br>";
    }

    function repartirCartas($cartasPedidas) {
        $cartasDisponibles = self::$NUM_CARTAS - $this->posSiguienteCarta;
        if ($cartasDisponibles > 0 && $cartasPedidas > 0 && $cartasDisponibles > $cartasPedidas) {
            for ($i = 0; $i < 4; $i++) {
                $this->siguienteCarta();
            }
        } else {
            return "No se pueden repartir las cartas";
        }
    }

    function cartasRepartidas() {
        for ($i = 0; $i < $this->posSiguienteCarta; $i++) {
            echo $this->cartas[$i]->toString();
        }
    }

    function mostrarBaraja() {
        echo "Mostrando baraja: INICIO <br>";
        for ($i = $this->posSiguienteCarta; $i < count($this->cartas); $i++) {
            echo $this->cartas[$i]->toString();
        }
        echo "Mostrando baraja: FIN <br><br>";
    }

}
